component/newstorage/batchedstore/batchedstore.go

```diff
@@ -0,0 +1,361 @@
+}
+
+// SetStoreConfig sets the configuration on a store.
+// The store must be created prior to calling this method.
+// If the store cannot be found, then an error wrapping ErrStoreNotFound will be returned by the underlying provider.
+// If name is blank, then an error will be returned by the underlying provider.
+func (p *Provider) SetStoreConfig(name string, config newstorage.StoreConfiguration) error {
+	err := p.underlyingProvider.SetStoreConfig(name, config)
+	if err != nil {
+		return fmt.Errorf("failed to set store config in underlying provider: %w", err)
+	}
+
+	return nil
+}
+
+// GetStoreConfig gets the current store configuration.
+// The store must be created prior to calling this method.
+// If the store cannot be found, then an error wrapping ErrStoreNotFound will be returned by the underlying provider.
+// If name is blank, then an error will be returned by the underlying provider.
+func (p *Provider) GetStoreConfig(name string) (newstorage.StoreConfiguration, error) {
+	config, err := p.underlyingProvider.GetStoreConfig(name)
+	if err != nil {
+		return newstorage.StoreConfiguration{},
+			fmt.Errorf("failed to get store config from underlying provider: %w", err)
+	}
+
+	return config, nil
+}
+
+// GetOpenStores returns all currently open stores.
+func (p *Provider) GetOpenStores() []newstorage.Store {
+	openStores := make([]newstorage.Store, len(p.openStores))
+
+	var counter int
+
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	for _, openStore := range p.openStores {
+		openStores[counter] = openStore
+		counter++
+	}
+
+	return openStores
+}
+
+// Close closes all stores created under this store provider.
+// For persistent store implementations, this does not delete any data in the underlying databases.
+func (p *Provider) Close() error {
+	var openStoresSnapshot []*store
+
+	p.lock.Lock()
+	for _, openStore := range p.openStores {
+		openStoresSnapshot = append(openStoresSnapshot, openStore)
+	}
+	p.lock.Unlock()
+
+	for _, openStore := range openStoresSnapshot {
+		err := openStore.Close()
+		if err != nil {
+			return fmt.Errorf(`failed to close open store with name "%s": %w`, openStore.name, err)
+		}
+	}
+
+	err := p.underlyingProvider.Close()
+	if err != nil {
+		return fmt.Errorf("failed to close underlying provider: %w", err)
+	}
+
+	return nil
+}
+
+func (p *Provider) removeStore(name string) {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	delete(p.openStores, name)
+}
+
+type store struct {
+	name            string
+	underlyingStore newstorage.Store
+	currentBatch    []newstorage.Operation
+	batchSizeLimit  int
+	close           closer
+	*sync.RWMutex
+}
+
+func (s *store) Put(key string, value []byte, tags ...newstorage.Tag) error {
+	if key == "" {
+		return errEmptyKey
+	}
+
+	if value == nil {
+		return errors.New("value cannot be nil")
+	}
+
+	s.Lock()
+
+	s.currentBatch = append(s.currentBatch, newstorage.Operation{
+		Key:   key,
+		Value: value,
+		Tags:  tags,
+	})
+
+	s.Unlock()
+
+	s.RLock()
```

Comment:
Don't need the extra read lock - just move the assignment inside the write lock above.

Reply:
Done